### PR TITLE
Add MakoTV extractor

### PR DESCRIPTION
**Pull request by relrelb**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Adding a new extractor for [MakoTV](https://www.mako.co.il/mako-vod), which is a free Israeli VOD service containing TV shows and more.
It has a flash player for displaying videos, and I managed to extract their m3u8 streams.
Seems like youtube-dl uses ffmpeg to download these streams, and it takes pretty long time. I would like to know if there's a better way of accomplishing that, and any other notes in general.

### CODE REVIEW

**Comment by dstftw** [comment]
Now actually read what's linked in boxes you've checked and fix all issues.

**Comment by relrelb** [comment]
Can you please be more specific?

**Comment by relrelb** [comment]
Are there any pending fixes to be done? If so, I'll fix them.

File: youtube_dl/extractor/makotv.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+try:
+    from urllib import urlencode
+    from urlparse import urlparse, urlunparse, urljoin
+except ImportError:
+    from urllib.parse import urlencode, urlparse, urlunparse, urljoin
```
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: `compat` module.

File: youtube_dl/extractor/makotv.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        parsed_url = list(urlparse(url))
+        parsed_url[4] = urlencode({'type': 'service'})
+        service = self._download_json(urlunparse(parsed_url), video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: `update_url_query`.

File: youtube_dl/extractor/makotv.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        config_new = self._download_xml('https://rcs.mako.co.il/flash_swf/players/makoPlayer/configNew.xml', video_id)
+        playlist_url = config_new.findtext('./PlaylistUrl')
+        playlist_url = playlist_url.replace('$$vcmid$$', video['guid'])
+        playlist_url = playlist_url.replace('$$videoChannelId$$', video['chId'])
+        playlist_url = playlist_url.replace('$$galleryChannelId$$', video['galleryChId'])
+        playlist = self._download_json('https://www.mako.co.il' + playlist_url, video_id)
+
+        formats = []
+        for media in playlist['media']:
+            tickets = self._download_json('https://mass.mako.co.il/ClicksStatistics/entitlementsServicesV2.jsp?et=gt&lp={}&rv={}'.format(media['url'], media['cdn']), video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: No `{}`.

Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: Query to `query=`.

Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: Should not be fatal.

File: youtube_dl/extractor/makotv.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            assert tickets['status'] == 'success'
+            for ticket in tickets['tickets']:
+                ticket_url = urljoin('https://makostore-hd.ctedgecdn.net', '{}?{}'.format(ticket['url'], ticket['ticket']))
+                formats.extend(self._extract_m3u8_formats(ticket_url, video_id))
```
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: Should not be fatal.

File: youtube_dl/extractor/makotv.py (hunk 3 above)
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: Should be tolerate to missing keys in `media`.

File: youtube_dl/extractor/makotv.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        playlist_url = playlist_url.replace('$$vcmid$$', video['guid'])
+        playlist_url = playlist_url.replace('$$videoChannelId$$', video['chId'])
+        playlist_url = playlist_url.replace('$$galleryChannelId$$', video['galleryChId'])
+        playlist = self._download_json('https://www.mako.co.il' + playlist_url, video_id)
+
+        formats = []
+        for media in playlist['media']:
+            tickets = self._download_json('https://mass.mako.co.il/ClicksStatistics/entitlementsServicesV2.jsp?et=gt&lp={}&rv={}'.format(media['url'], media['cdn']), video_id)
+            assert tickets['status'] == 'success'
+            for ticket in tickets['tickets']:
```
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: Same tolerance issue.

**Comment by relrelb** [comment]
Thanks for the review, I think I fixed the issues you mentioned.
By the way, why the download shows progress of ffmpeg? How can I omit it?

**Comment by relrelb** [comment]
Is it still going?

**Comment by dstftw** [comment]
For any work on this you must provide account credentials.

File: youtube_dl/extractor/makotv.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            'upload_date': '20120708',
+            'timestamp': 1341751140
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        service = self._download_json(update_url_query(url, {'type': 'service'}), video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: Query to `query`.

File: youtube_dl/extractor/makotv.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+class MakoTVIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?mako\.co\.il/mako-vod-.+/VOD-(?P<id>[0-9a-f]{18})\.htm'
+    _TEST = {
+        'url': 'https://www.mako.co.il/mako-vod-keshet/parliament-s1/VOD-5df5a86c1966831006.htm',
+        'md5': 'cd8cdff75390f8521831ec5049841764',
+        'info_dict': {
+            'id': '5df5a86c1966831006',
+            'ext': 'm3u8',
+            'title': 'הפרלמנט | פרק 1 לצפייה ישירה | makoTV ',
```
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: Must only contain title.

File: youtube_dl/extractor/makotv.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            'thumbnail': r're:^https?://img\.mako\.co\.il/\d{4}/\d{2}/\d{2}/.*\.jpg$',
+            'description': 'שאולי, אמציה, הקטור, קרקו ואבי מקבלים סדרה משלהם. כל הפרקים של הפרלמנט לצפייה ישירה | makoTV ',
```
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: Must only contain description.

File: youtube_dl/extractor/makotv.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        config_new = self._download_xml('https://rcs.mako.co.il/flash_swf/players/makoPlayer/configNew.xml', video_id)
+        playlist_url = config_new.findtext('./PlaylistUrl')
+        playlist_url = playlist_url.replace('$$vcmid$$', video['guid'])
+        playlist_url = playlist_url.replace('$$videoChannelId$$', video['chId'])
+        playlist_url = playlist_url.replace('$$galleryChannelId$$', video['galleryChId'])
+        playlist = self._download_json(urljoin('https://www.mako.co.il', playlist_url), video_id)
+
+        formats = []
+        for media in playlist.get('media', []):
+            tickets = self._download_json('https://mass.mako.co.il/ClicksStatistics/entitlementsServicesV2.jsp', video_id, fatal=False, query={'et': 'gt', 'lp': media['url'], 'rv': media['cdn']})
```
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: Carry long lines. Read coding conventions.

File: youtube_dl/extractor/makotv.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            if tickets is None or tickets.get('status') != 'success':
+                continue
+            for ticket in tickets.get('tickets', {}):
+                ticket_url = urljoin('https://makostore-hd.ctedgecdn.net', ticket['url']) + "?" + ticket['ticket']
+                formats.extend(self._extract_m3u8_formats(ticket_url, video_id, fatal=False))
+
+        self._sort_formats(formats)
+
+        info = self._search_json_ld(webpage, video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/makotv.py]: Must not be fatal.

File: youtube_dl/extractor/makotv.py (hunk 7 above)
Comment (relrelb) [reply youtube_dl/extractor/makotv.py]: This is how the title appears in the `<title>` tag, in the `content` of the `<meta property="og:title">` tag, and in the `"name"` of the `<script type="application/ld+json">` tag.
I think the best I can do is to remove the "makoTV" suffix with a regex or something.
What do you think?

**Comment by relrelb** [comment]
By the way, as the extractor downloads .m3u8, it prints a lot of verbose logs of ffmpeg. Is there any way to silent those logs and/or replace them with a nicer progress instead?

**Comment by dstftw** [comment]
Not possible.

**Comment by relrelb** [comment]
I noticed [this pull request](https://github.com/ytdl-org/youtube-dl/pull/17997) that adds Mako too. So I took inspiration from it and made this version more generic.

**Comment by relrelb** [comment]
@dstftw Can we merge this PR?

**Comment by relrelb** [comment]
@dstftw Can we merge this?

**Pull request closed without merging**